pkg/util/resources.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"strings"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+)
+
+// ParseResourcesString takes a comma separated string list of resources in the form of
+// <api-group1>, <api-group2> .. and returns a map[string]bool
+func ParseResourceGroupsString(resourceGroups string) map[string]bool {
+	if resourceGroups == "" {
+		return nil
+	}
+
+	// trim single and double quotes to make this safe from
+	// user passing the resource option quoted in the container args
+	resourceGroups = strings.Trim(resourceGroups, `'"`)
+
+	groupsMap := map[string]bool{}
+	for _, g := range strings.Split(resourceGroups, ",") {
+		groupsMap[strings.Trim(g, " ")] = true
+	}
+
+	return groupsMap
+}
+
+// IsResourceGroupAllowed checks if an infomer for the resource is allowed to start.
+// an empty or nil allowedResources slice is equivalent to allow all,
+func IsResourceGroupAllowed(resourceGroup string, allowedResourceGroups map[string]bool) bool {
+	if allowedResourceGroups == nil || allowedResourceGroups != nil && len(allowedResourceGroups) == 0 {
+		return true
+	}
+	addRequiredResourceGroups(allowedResourceGroups)
+
+	return allowedResourceGroups[resourceGroup]
+}
+
+// append the minimal set of resources that are required to operate
+func addRequiredResourceGroups(allowedResourceGroups map[string]bool) {
+	// if resources are provided, we need to ensure that at least CRD and KS API
+	// resources are watched
+
+	allowedResourceGroups[v1alpha1.GroupVersion.Group] = true
+
+	// disabled until https://github.com/kubestellar/kubestellar/issues/1705 is resolved
+	// to avoid client-side throttling
```

Comment:
I do not understand this. #1705 is about a high request rate from this controller for discovery. I do not see how disallowing that one API group makes a difference --- the full discovery is going to be done anyway, right? And even if only allowed groups were queried in discovery, adding this one API group would not greatly increase the requests for discovery.

Reply:
It's not just that one extra informer for CRD is enabled. When that informer is enabled, https://github.com/kubestellar/kubestellar/blob/a357f6b0783627fd90ff883ad4716d386e8ee4dc/pkg/placement/crd.go#L71 is invoked at startup for each CRD on the system (and there are ~ 300). The call to https://github.com/kubestellar/kubestellar/blob/a357f6b0783627fd90ff883ad4716d386e8ee4dc/pkg/placement/crd.go#L82 was generating a lot of client-side throttling. This issue only shows up when you have a large number of CRDs. I have opened an issue to address that in a different PR.

Reply:
So the proposal here is to break everything involving CRDs and fix it in follow-on?

Reply:
yes, but only for this specific case of restricting api groups, because this is what was needed for the current test with MCAD and OCP as hosting cluster.

Reply:
If I understand correctly, this PR stops the controller from properly reacting to CRD add/change/delete but does not stop the controller from handling resources defined by CRDs that were fully in effect before the controller started. Is that right? And you are saying that the plan for the MCAD stack is to get the MCAD CRDs fully in effect before launching this controller. Is that right?

Reply:
that's right, on both counts. MCAD CRD is installed on host cluster (and wecs) before starting the controller.